File: artificialanalysis.ai-parser.js
/**
 * artificialanalysis.ai-parser.js
 * 
 * Node.js parser for artificialanalysis.ai model data.
 * Extracts pricing, benchmarks, and speed from the public RSC endpoint
 * — no API key required.
 * 
 * ⚠️ BROWSER LIMITATION: The RSC endpoint requires the custom 'rsc' header
 * which triggers a CORS preflight. The server does NOT return
 * Access-Control-Allow-Headers, so browser fetch() is blocked.
 * Use in Node.js or through a CORS proxy.
 * 
 * Usage (Node.js):
 *   const { AAParser } = require('./artificialanalysis.ai-parser.js');
 *   const models = await AAParser.fetch();
 *   console.log(models[0].name, models[0].price_1m_input_tokens);
 * 
 * Usage (browser — requires CORS proxy):
 *   <script src="artificialanalysis.ai-parser.js"></script>
 *   <script>
 *     // Only works with a CORS proxy that strips/forwards custom headers
 *     const models = await AAParser.fetch();
 *   </script>
 * 
 * License: GPL-3.0 — Copyright (C) 2026 Anton Maurer
 */

(function (root, factory) {
    if (typeof define === 'function' && define.amd) {
        define([], factory);
    } else if (typeof module === 'object' && module.exports) {
        module.exports = factory();
    } else {
        root.AAParser = factory();
    }
}(typeof self !== 'undefined' ? self : this, function () {

    'use strict';

    var RSC_URL = 'https://artificialanalysis.ai/leaderboards/providers?_rsc=hgvan';
    var RSC_HEADERS = {
        'accept': '*/*',
        'rsc': '1',
        'next-router-prefetch': '1',
        'next-router-state-tree': '[["","pages",["leaderboards",["models",["__PAGE__",{},"/leaderboards/models","refresh"]]]],null,null,true]',
        'next-url': '/leaderboards/models'
    };

    /**
     * Download the RSC stream as raw text.
     * @returns {Promise<string>}
     */
    function fetchRSC() {
        if (typeof fetch === 'undefined') {
            return Promise.reject(new Error('fetch() not available — use Node.js 18+ or a browser'));
        }
        return fetch(RSC_URL, { headers: RSC_HEADERS })
            .then(function (resp) {
                if (!resp.ok) throw new Error('HTTP ' + resp.status);
                return resp.text();
            });
    }

    /**
     * Extract the 'hostsModels' JSON array from the RSC stream.
     * @param {string} raw
     * @returns {Array<Object>}
     */
    function extractHostsModels(raw) {
        var idx = raw.indexOf('"hostsModels"');
        if (idx === -1) throw new Error("'hostsModels' not found in response");

        var arrStart = raw.indexOf('[', idx);
        if (arrStart === -1 || arrStart - idx > 500)
            throw new Error("Could not find array start after 'hostsModels'");

        // Bracket-counting to find matching close bracket
        var depth = 1;
        var pos = arrStart + 1;
        while (depth > 0 && pos < raw.length) {
            var c = raw.charAt(pos);
            if (c === '[') depth++;
            else if (c === ']') depth--;
            if (depth > 0) pos++;
        }

        return JSON.parse(raw.slice(arrStart, pos + 1));
    }

    /**
     * Deduplicate by model_id, keeping first occurrence (full non-circular data).
     * @param {Array<Object>} entries
     * @returns {Array<Object>}
     */
    function deduplicate(entries) {
        var seen = {};
        var result = [];
        for (var i = 0; i < entries.length; i++) {
            var mid = entries[i].model_id;
            if (mid && !seen[mid]) {
                seen[mid] = true;
                result.push(entries[i]);
            }
        }
        return result;
    }

    /**
     * Clean a raw entry into a tidy model object.
     * @param {Object} entry
     * @returns {Object}
     */
    function cleanModel(entry) {
        var model = entry.model || {};
        var ts = entry.timescaleData || {};
        var host = entry.host || {};
        var creator = (model.model_creators) || {};

        return {
            name: model.name || entry.model_label || '?',
            creator: creator.name || host.name || '?',
            provider: host.name || '?',
            slug: model.slug || '',
            intelligence_index: model.intelligence_index != null ? model.intelligence_index : null,
            coding_index: model.coding_index != null ? model.coding_index : null,
            math_index: model.math_index != null ? model.math_index : null,
            agentic_index: model.agentic_index != null ? model.agentic_index : null,
            price_1m_input_tokens: entry.price_1m_input_tokens != null ? entry.price_1m_input_tokens : null,
            price_1m_output_tokens: entry.price_1m_output_tokens != null ? entry.price_1m_output_tokens : null,
            price_1m_cache_hit: entry.cache_hit_price != null ? entry.cache_hit_price : null,
            blended_price_3_1: entry.price_1m_blended_3_1 != null ? entry.price_1m_blended_3_1 : null,
            context_window_tokens: entry.context_window_tokens != null ? entry.context_window_tokens : null,
            output_tokens_per_second: ts.median_output_speed != null ? ts.median_output_speed : null,
            time_to_first_token_ms: ts.median_time_to_first_token != null
                ? Math.round(ts.median_time_to_first_token * 1000 * 10) / 10
                : null,
            reasoning: !!model.reasoning_model,
            open_weights: !!model.is_open_weights,
            release_date: model.release_date || null,
            gpqa: model.gpqa != null ? model.gpqa : null,
            mmlu_pro: model.mmlu_pro != null ? model.mmlu_pro : null,
            hle: model.hle != null ? model.hle : null
        };
    }

    /**
     * Strip to minimal fields for the cost calculator.
     * @param {Array<Object>} models
     * @returns {Array<Object>}
     */
    function compress(models) {
        var out = [];
        for (var i = 0; i < models.length; i++) {
            var m = models[i];
            out.push({
                name: m.name,
                creator: m.creator,
                slug: m.slug,
                intelligence_index: m.intelligence_index,
                coding_index: m.coding_index,
                math_index: m.math_index,
                price_1m_input_tokens: m.price_1m_input_tokens,
                price_1m_output_tokens: m.price_1m_output_tokens,
                price_1m_cache_hit: m.price_1m_cache_hit,
                blended_price_3_1: m.blended_price_3_1,
                context_window_tokens: m.context_window_tokens,
                output_tokens_per_second: m.output_tokens_per_second,
                time_to_first_token_ms: m.time_to_first_token_ms,
                reasoning: m.reasoning,
                open_weights: m.open_weights
            });
        }
        return out;
    }

    /**
     * Fetch, parse, deduplicate, and clean models from artificialanalysis.ai.
     * 
     * @param {Object} [opts]
     * @param {boolean} [opts.minimal=false] — return only calculator-essential fields
     * @param {string} [opts.raw] — parse an existing RSC text instead of fetching
     * @returns {Promise<Array<Object>>}
     */
    function fetchModels(opts) {
        opts = opts || {};
        var rawPromise;
        if (opts.raw) {
            rawPromise = Promise.resolve(opts.raw);
        } else {
            rawPromise = fetchRSC();
        }

        return rawPromise.then(function (raw) {
            var entries = extractHostsModels(raw);
            var deduped = deduplicate(entries);
            var models = deduped.map(cleanModel);

            // Keep only entries with pricing
            models = models.filter(function (m) {
                return m.price_1m_input_tokens != null && m.price_1m_output_tokens != null;
            });

            // Sort by intelligence index descending
            models.sort(function (a, b) {
                return (b.intelligence_index || 0) - (a.intelligence_index || 0);
            });

            if (opts.minimal) {
                models = compress(models);
            }

            return models;
        });
    }

    // ── Public API ──
    return {
        fetch: fetchModels,
        extractHostsModels: extractHostsModels,
        deduplicate: deduplicate,
        cleanModel: cleanModel,
        compress: compress,
        VERSION: '1.0.0'
    };

}));

File: artificialanalysis.ai-parser.py
#!/usr/bin/env python3
"""
fetch_aa.py — Extract model data from artificialanalysis.ai RSC endpoint
No API key required. Outputs clean JSON for the cost calculator.

Usage:
    python3 fetch_aa.py                    # fetch + parse, save to models.json
    python3 fetch_aa.py --file aa.txt      # parse existing RSC dump
    python3 fetch_aa.py --out models.json  # custom output path
    python3 fetch_aa.py --minimal          # only fields needed for calculator
"""

import json
import sys
import os
import argparse
import time
from urllib.request import Request, urlopen
from urllib.error import URLError

RSC_URL = "https://artificialanalysis.ai/leaderboards/providers?_rsc=hgvan"
RSC_HEADERS = {
    "accept": "*/*",
    "rsc": "1",
    "next-router-prefetch": "1",
    "next-router-state-tree": '[["","pages",["leaderboards",["models",["__PAGE__",{},"/leaderboards/models","refresh"]]]],null,null,true]',
    "next-url": "/leaderboards/models",
    "user-agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/138.0.0.0 Safari/537.36",
}


def fetch_rsc(timeout=60):
    """Download the RSC stream from artificialanalysis.ai."""
    print(f"Downloading RSC data from {RSC_URL} ...")
    req = Request(RSC_URL, headers=RSC_HEADERS)
    try:
        with urlopen(req, timeout=timeout) as resp:
            if resp.status != 200:
                print(f"Error: HTTP {resp.status}")
                return None
            raw = resp.read()
            print(f"Downloaded {len(raw):,} bytes")
            return raw
    except URLError as e:
        print(f"Network error: {e}")
        return None


def extract_hosts_models(raw):
    """Extract the 'hostsModels' JSON array from the RSC stream."""
    idx = raw.find(b'"hostsModels"')
    if idx < 0:
        print("Error: 'hostsModels' not found in response")
        return None

    arr_start = raw.find(b'[', idx)
    if arr_start < 0 or arr_start - idx > 500:
        print("Error: could not find array start after 'hostsModels'")
        return None

    # Bracket-counting to find the matching close bracket
    depth = 1
    pos = arr_start + 1
    while depth > 0 and pos < len(raw):
        b = raw[pos]
        if b == 0x5B:     # [
            depth += 1
        elif b == 0x5D:   # ]
            depth -= 1
        if depth > 0:
            pos += 1

    json_bytes = raw[arr_start:pos + 1]
    try:
        return json.loads(json_bytes)
    except json.JSONDecodeError as e:
        print(f"JSON parse error at byte {e.pos}: {e.msg}")
        return None


def deduplicate_models(entries):
    """Deduplicate by model_id, keeping first occurrence (has full data)."""
    seen = {}
    for entry in entries:
        mid = entry.get("model_id")
        if mid and mid not in seen:
            seen[mid] = entry
    return list(seen.values())


def clean_model(entry):
    """Extract clean model data from a raw entry."""
    model = entry.get("model", {})
    ts = entry.get("timescaleData", {}) or {}
    host = entry.get("host", {}) or {}
    creator = model.get("model_creators", {}) or {}

    return {
        "name": model.get("name", entry.get("model_label", "?")),
        "creator": creator.get("name", host.get("name", "?")),
        "provider": host.get("name", "?"),
        "slug": model.get("slug", ""),
        "intelligence_index": model.get("intelligence_index"),
        "coding_index": model.get("coding_index"),
        "math_index": model.get("math_index"),
        "agentic_index": model.get("agentic_index"),
        "price_1m_input_tokens": entry.get("price_1m_input_tokens"),
        "price_1m_output_tokens": entry.get("price_1m_output_tokens"),
        "price_1m_cache_hit": entry.get("cache_hit_price"),
        "blended_price_3_1": entry.get("price_1m_blended_3_1"),
        "blended_price_7_2_1": entry.get("price_1m_blended_7_2_1"),
        "context_window_tokens": entry.get("context_window_tokens"),
        "output_tokens_per_second": ts.get("median_output_speed"),
        "time_to_first_token_ms": (
            round(ts["median_time_to_first_token"] * 1000, 1)
            if ts.get("median_time_to_first_token") else None
        ),
        "reasoning": model.get("reasoning_model", False),
        "open_weights": model.get("is_open_weights", False),
        "release_date": model.get("release_date"),
        "gpqa": model.get("gpqa"),
        "mmlu_pro": model.get("mmlu_pro"),
        "hle": model.get("hle"),
    }


def compress_for_calculator(models):
    """Return minimal fields needed by the aiprice.html calculator."""
    result = []
    for m in models:
        result.append({
            "name": m["name"],
            "creator": m["creator"],
            "slug": m["slug"],
            "intelligence_index": m["intelligence_index"],
            "coding_index": m["coding_index"],
            "math_index": m["math_index"],
            "price_1m_input_tokens": m["price_1m_input_tokens"],
            "price_1m_output_tokens": m["price_1m_output_tokens"],
            "price_1m_cache_hit": m["price_1m_cache_hit"],
            "blended_price_3_1": m["blended_price_3_1"],
            "context_window_tokens": m["context_window_tokens"],
            "output_tokens_per_second": m["output_tokens_per_second"],
            "time_to_first_token_ms": m["time_to_first_token_ms"],
            "reasoning": m["reasoning"],
            "open_weights": m["open_weights"],
        })
    return result


def main():
    parser = argparse.ArgumentParser(description="Fetch AI model data from artificialanalysis.ai")
    parser.add_argument("--file", help="Parse existing RSC dump file (skip download)")
    parser.add_argument("--out", default="models.json", help="Output JSON file (default: models.json)")
    parser.add_argument("--minimal", action="store_true", help="Output only calculator-essential fields")
    parser.add_argument("--pretty", action="store_true", help="Pretty-print JSON output")
    args = parser.parse_args()

    # Step 1: Get raw data
    if args.file:
        if not os.path.exists(args.file):
            print(f"Error: file '{args.file}' not found")
            sys.exit(1)
        with open(args.file, "rb") as f:
            raw = f.read()
        print(f"Read {len(raw):,} bytes from {args.file}")
    else:
        raw = fetch_rsc()
        if raw is None:
            sys.exit(1)

    # Step 2: Extract hostsModels
    entries = extract_hosts_models(raw)
    if entries is None:
        sys.exit(1)
    print(f"Extracted {len(entries)} raw entries (host-model pairs)")

    # Step 3: Deduplicate
    deduped = deduplicate_models(entries)
    print(f"Deduplicated to {len(deduped)} unique models")

    # Step 4: Clean
    models = [clean_model(e) for e in deduped]

    # Remove entries without pricing
    models_with_price = [m for m in models if m["price_1m_input_tokens"] and m["price_1m_output_tokens"]]
    print(f"Models with pricing: {len(models_with_price)}")

    # Sort by intelligence index descending
    models_with_price.sort(key=lambda m: m["intelligence_index"] or 0, reverse=True)

    # Step 5: Output
    if args.minimal:
        output = compress_for_calculator(models_with_price)
    else:
        output = models_with_price

    indent = 2 if args.pretty else None
    with open(args.out, "w", encoding="utf-8") as f:
        json.dump(output, f, indent=indent, ensure_ascii=False)

    size = os.path.getsize(args.out)
    print(f"\nSaved {len(output)} models to {args.out} ({size:,} bytes)")

    # Stats
    if output:
        top = output[0]
        print(f"\nTop model: {top['name']} ({top['creator']})")
        print(f"  IQ: {top['intelligence_index']} | Coding: {top['coding_index']} | Math: {top['math_index']}")
        print(f"  Price: ${top['price_1m_input_tokens']:.2f} in / ${top['price_1m_output_tokens']:.2f} out")
        if top.get("output_tokens_per_second"):
            print(f"  Speed: {top['output_tokens_per_second']:.0f} tok/s")


if __name__ == "__main__":
    main()
